common/src/prover_state/mod.rs

```diff
@@ -229,13 +231,26 @@ impl ProverStateManager {
         match self.persistence {
             CircuitPersistence::None | CircuitPersistence::Disk(TableLoadStrategy::Monolithic) => {
                 info!("using monolithic circuit {:?}", self);
-                self.txn_proof_monolithic(input)
+                self.txn_proof_monolithic(input.clone())
             }
             CircuitPersistence::Disk(TableLoadStrategy::OnDemand) => {
                 info!("using on demand circuit {:?}", self);
-                self.txn_proof_on_demand(input)
+                self.txn_proof_on_demand(input.clone())
             }
         }
+        .map_err(|e| {
+            if log_enabled!(log::Level::Debug) {
+                if let Err(write_err) = save_inputs_to_disk(
+                    format!("b{}_txn_input.log", input.block_metadata.block_number),
```

Comment:
If we have 10 workers in parallel, and 2 fail, the secondly failing txn will overwrite the first file here (because they aren't distinguished past the block number). Probably better to have something like:

```suggestion
                    format!("b{}_txn_{}_input.log", input.block_metadata.block_number, input.txn_number_before),
```

for instance